testing/trino-testing/src/main/java/io/trino/testing/AbstractTestFailureRecovery.java

```diff
@@ -0,0 +1,626 @@
+            return this;
+        }
+
+        private ExecutionResult executeExpected()
+        {
+            return execute(query, Optional.empty());
+        }
+
+        private ExecutionResult executeActual(MaterializedResult expected)
+        {
+            requireNonNull(stageSelector, "stageSelector must be set");
+            int stageId = stageSelector.apply(expected);
+            String token = UUID.randomUUID().toString();
+            failureType.ifPresent(failure -> queryRunner.injectTaskFailure(
+                    token,
+                    stageId,
+                    0,
+                    0,
+                    failure,
+                    errorType));
+
+            ExecutionResult actual = execute(query, Optional.of(token));
+            assertEquals(getStageStats(actual.getQueryResult(), stageId).getFailedTasks(), failureType.isPresent() ? 1 : 0);
+            return actual;
+        }
+
+        private ExecutionResult execute(String query, Optional<String> traceToken)
+        {
+            String tableName = "table_" + randomTableSuffix();
+            setup.ifPresent(sql -> queryRunner.execute(session, resolveTableName(sql, tableName)));
+
+            MaterializedResult result = null;
+            RuntimeException failure = null;
+            try {
+                Session sessionWithToken = Session.builder(session)
+                        .setTraceToken(traceToken)
+                        .build();
+                result = queryRunner.execute(sessionWithToken, resolveTableName(query, tableName));
+            }
+            catch (RuntimeException e) {
+                failure = e;
+            }
+
+            Optional<MaterializedResult> updatedTableContent = Optional.empty();
+            if (result != null && result.getUpdateCount().isPresent()) {
+                updatedTableContent = Optional.of(queryRunner.execute(session, "SELECT * FROM " + tableName));
+            }
+
+            Optional<MaterializedResult> updatedTableStatistics = Optional.empty();
+            if (result != null && result.getUpdateType().isPresent() && result.getUpdateType().get().equals("ANALYZE")) {
+                updatedTableStatistics = Optional.of(queryRunner.execute(session, "SHOW STATS FOR " + tableName));
+            }
+
+            try {
+                cleanup.ifPresent(sql -> queryRunner.execute(session, resolveTableName(sql, tableName)));
+            }
+            catch (RuntimeException e) {
+                if (failure == null) {
+                    failure = e;
+                }
+                else if (failure != e) {
+                    failure.addSuppressed(e);
+                }
+            }
+
+            if (failure != null) {
+                throw failure;
+            }
+
+            return new ExecutionResult(result, updatedTableContent, updatedTableStatistics);
+        }
+
+        public void finishesSuccessfully()
```

Comment:
Can you add an assertion here that retries actually happened. (maybe method should be called `finishesSuccessfullyWithRetries`)

Reply:
The `executeActual` already does that. It verifies query statistics to confirm there is a failed task.